fairscale/experimental/nn/mevo.py

```diff
@@ -0,0 +1,483 @@
+            # This means we accumulate full grad between iters. Not memory efficient.
+            ctx.kernel_obj.proj_weight.grad.add_(w.grad)
+        else:
+            ctx.kernel_obj.proj_weight.grad = w.grad
+        return i.grad, None, None, None
+
+
+class BackwardTriggerFn(torch.autograd.Function):
+    """A backward trigger function."""
+
+    @staticmethod
+    def forward(  # type: ignore
+        ctx: Any, w: torch.Tensor, trigger_tensor: torch.Tensor
+    ) -> torch.Tensor:
+        """We take a weight tensor and the trigger as inputs and output the weight directly."""
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            print("DEBUG trigger fwd")
+        ctx.save_for_backward(w, trigger_tensor)
+        return w
+
+    @staticmethod
+    def backward(ctx: Any, *args: Any) -> Any:
+        """We return zero grad for the trigger only."""
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            print("DEBUG trigger bwd")
+        assert len(args) == 1
+        w, trigger = ctx.saved_tensors
+        assert w.requires_grad
+        assert trigger.requires_grad
+        return None, torch.zeros_like(trigger)
+
+
+class BackwardTrigger(nn.Module):
+    """A backward trigger module.
+
+       This module takes a parameter as an input and create a linked parameter
+       from a newly created trigger parameter.
+
+       The way to use it in a module's ``__init__'' and ``forward'' functions:
+
+       ```
+       def __init__():
+         ...
+         self.trigger = BackwardTrigger(some_layer.weight)
+         ...
+
+       def forward():
+         w = self.trigger()
+         ... continue to use w ...
+       ```
+
+       As a resule, the trigger's backward hook will be called at the end of
+       the backward for the module that uses this trigger.
+    """
+
+    def __init__(self, linked_param: torch.Tensor):
+        super().__init__()
+        assert isinstance(linked_param, nn.Parameter)
+        self.trigger = nn.Parameter(torch.rand(1, dtype=linked_param.dtype))
+        self.trigger._linked_param = linked_param
+
+    def forward(self) -> torch.Tensor:  # type: ignore
+        return BackwardTriggerFn.apply(self.trigger._linked_param, self.trigger)
+
+
+class MemoryEfficientVocabOutput(nn.Module):  # AKA. MEVO
+    """ Fused fc + softmax + nll_loss in a tiled fashion.
+
+        This uses much less memory but is quite a bit slower.
+
+    Args:
+        proj_weight (nn.Parameter):
+            Sharing this weight with an embedding layer.
+        tile_factor (int):
+            Number of splits to use on the input sequence and vocab dimensions.
+        reduction (str):
+            Reduction OP (sum or mean).
+    """
+
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 16, reduction: str = "sum"):
+        super().__init__()
+        self.proj_weight = proj_weight
+        self.tf_in, self.tf_w = tile_factor, tile_factor
```

Comment:
does tile_factor have to be the same? how does it impact memory savings?

Reply:
It doesn't need to be the same. I haven't extensively tuning it. Will add a TODO.